docs/examples/profiling.ipynb

```diff
@@ -0,0 +1,622 @@
+    "- RAM/VRAM utilization\n",
+    "\n",
+    "In the Mila Research Template, this can be done by passing a callback to the trainer. Supported configs are found within the project template at `configs/trainer/callbacks`. Throughout this tutorial, we will use the default callback, which in turn implements early stopping and tracks the learning rate, device utilisation and throughput, each through a specific callback instance."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Running a baseline and logging metrics on WandB"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In addition to specifying callbacks, the Mila Research template integrates using WandB as a logger, which enables the tracking of additional metrics through visualizations and dashboard creation. Given the flexibility and widespread adoption of the WandB logger, we'll be using it for the remainder of this tutorial, which will then be visualizable in the following [WandB report](https://wandb.ai/cesar-valdez-mcgill-university/ResearchTemplate/reports/Profiling--Vmlldzo5NDI1MjU0) , as supporting information for the experiments contained herein. We will now proceed to establish a baseline to profile, diagnose and optimize throughout this example."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "!python project/main.py \\\n",
+    "    experiment=profiling \\\n",
+    "    trainer.logger.wandb.name=\"Baseline with training\" \\\n",
+    "    trainer.logger.wandb.tags=[\"Training\",\"Baseline comparison\", \"CPU/GPU comparison\"] "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Identifying potential bottlenecks "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finding bottlenecks in your code is not necessarily clear or straightforward from the start. A sensible first step is to determine whether potential slowdowns originate from data loading or model computation. Running a model with and without training and contrasting the obtained outputs can help us determine whether the master process has a significant stall when fetching the next batch for training or not. Analyzing the difference between outputs can tell us the following about our model: \n",
+    "\n",
+    "- If the difference between data loading and training is close to 0, then the data loading procedure outpaces model computation, thus data loading is the bottleneck.\n",
+    "- If the difference between data loading and training is much greater than 0, then model computation outpaces data loading, thus computation is the bottleneck. \n",
+    "\n",
+    "We will proceed to run a series of experiments to identify potential bottlenecks: changing the workers involved in the dataloading process and the numbers of cpu assigned per task when training on a GPU."
```

Comment:
```suggestion
    "The first potential bottleneck to look out for is dataloading. An easy first step is to measure the throughput of your dataloading pipeline without any training. This can easily be done in this template with the `no_op` algorithm, which simply pulls batches from the dataloader, without doing any computation.  By Comparing the throughput (in samples/sec) that we get with the no-op algorithm vs our algorithm, we can infer the following: \n",
    "\n",
    "- If the throughput is much higher without training (e.g. >3x faster), then the slowest part of our code is the model computation. This is good.\n",
    "- If the difference in throughput (samples per second) between with and without training isn't significant, then data loading is the bottleneck, and we know where to focus our efforts to make the code more better. \n",
    "\n",
    "As can be seen in this figure (TODO: Add figure), we fall in the second case here: our dataloader is our bottleneck. Next, we will demonstrate how to improve the dataloading performance by changing the number of  workers and the numbers of cpus."
```